crates/scmi/src/devices/common.rs

```diff
@@ -0,0 +1,550 @@
+// SPDX-FileCopyrightText: Red Hat, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Common functionality for SCMI bindings to host devices.
+//!
+//! A new kind of devices can be added in [available_devices] using
+//! [DeviceSpecification::new] calls.
+//!
+//! The module also defines common infrastructure to provide sensor devices to
+//! SCMI, see [SensorT].
+
+use itertools::Itertools;
+use log::debug;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Write;
+use std::{format, println, writeln};
+
+use crate::scmi::{
+    self, DeviceResult, MessageId, MessageValue, MessageValues, ProtocolId, ScmiDevice,
+    ScmiDeviceError, MAX_SIMPLE_STRING_LENGTH, SENSOR_AXIS_DESCRIPTION_GET, SENSOR_CONFIG_GET,
+    SENSOR_CONFIG_SET, SENSOR_CONTINUOUS_UPDATE_NOTIFY, SENSOR_DESCRIPTION_GET, SENSOR_PROTOCOL_ID,
+    SENSOR_READING_GET,
+};
+
+use super::{fake, iio};
+
+// [(NAME, [(PROPERTY, VALUE), ...]), ...]
+pub(crate) type DeviceDescription = Vec<(String, DeviceProperties)>;
+type PropertyPairs = Vec<(String, String)>;
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub(crate) struct DeviceProperties(PropertyPairs);
+
+impl DeviceProperties {
+    pub(crate) fn new(properties: PropertyPairs) -> DeviceProperties {
+        DeviceProperties(properties)
+    }
+
+    pub(crate) fn get(&self, name: &str) -> Option<&String> {
+        self.0.iter().find(|(n, _)| n == name).map(|(_, v)| v)
+    }
+
+    fn names(&self) -> HashSet<&String> {
```

Comment:
I find it weird that some methods are returning `HashSet<&String>` & others `HashSet<String>`